### PR TITLE
chore: replace hashicorp/vault with openbao

### PR DESCRIPTION
**Pull request by mschfh**
### Summary

This PR removes `github.com/hashicorp/vault` to reduce transitive dependencies, such as `github.com/Azure/azure-sdk-for-go`.

before:
```
1.5 GiB [##########################] /github.com
386.7 MiB [######                    ] /google.golang.org
 78.7 MiB [#                         ] /golang.org
 20.3 MiB [                          ] /go.mongodb.org
```

after:
```
480.8 MiB [##########################] /github.com
385.6 MiB [####################      ] /google.golang.org
 78.8 MiB [####                      ] /golang.org
 15.6 MiB [                          ] /cloud.google.com
```

Fixes https://github.com/brave-intl/bat-go/security/dependabot/143.
Fixes https://github.com/brave-intl/bat-go/security/dependabot/144.
Fixes https://github.com/brave-intl/bat-go/security/dependabot/145.
Fixes https://github.com/brave-intl/bat-go/security/dependabot/146.
Fixes https://github.com/brave-intl/bat-go/security/dependabot/147.
Fixes https://github.com/brave-intl/bat-go/security/dependabot/148.

### Type of Change

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [X] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -67,6 +68,21 @@ jobs:
         go-version: 1.22
         cache-dependency-path: "**/go.sum"
 
+    # The runner has a rather small / file system, and tests are failing with "no space left on device" error.
+    # This frees up the space by deleteting uninstalling not needed packages.
+    - name: Free Disk Space
```
Comment (pavelbrm) [inline .github/workflows/ci.yml:73]: Note: this frees 5gb of space.

**Review by pavelbrm — APPROVED** [review]

**Comment by kdenhartog** [comment]
I've looked through this and not concerned with switching over to OpenBAO nor with the modifications made to Shamir.

I've also opened an [issue](https://github.com/brave-intl/bat-go/issues/2708) to remove the payments service from this repo since we've now migrated it.

**Comment by cipherboy** [comment]
\o hey @pavelbrm @kdenhartog -- quick question for y'all since I see you pulled the OpenBao SDK --- My understanding was that our `replace` directive would [make it hard for users to consume](https://github.com/openbao/openbao/pull/705#pullrequestreview-2435486904), but it doesn't seem like that's the case?

**Comment by pavelbrm** [comment]
Hey @cipherboy thanks for reaching out!

> our replace directive would make it hard for users to consume

It depends on how diligently the code base will be maintained. Any features from the replaced modules that are available via a released version should also be available to consumers. This means that when the parent module (that which contains the `replace` directives) is ready for releasing (creating a new tag), the inner module (the replaced one the outer code depends on) should be tagged and released first.

As far as I can tell, the original repository, `hashicorp/vault` does not do a good job at that.

To illustrate what I am referring to, consider the following example from the original repo. It replaces many of its internal components, `api` among them, [here](https://github.com/hashicorp/vault/blob/main/go.mod#L15). So everywhere in that module they rely on the most recent code.

But the most recent publicly available code for `api` is [v1.15.0](https://github.com/hashicorp/vault/releases/tag/api%2Fv1.15.0). At some point, yet unpublished features from the `api` module had been used throughout the parent module, and it resulted in problems.

I noticed an issue whilst upgrading `vault` from `v1.13.12` to `v1.18.1`. The latter uses code from the unpublished version of `api` in some parts, while others still depend on `api/v.1.15.0`. To fix the problem, I had to find out the commit of `v1.18.1`, and then upgrade `api` to that commit so that dependencies could be met (the build was failing because `api/v1.15.0` was lacking some fields that were added and used by `vault/v1.18.1`).

Ultimately, during the work on [that PR](https://github.com/brave-intl/bat-go/pull/2662), it was decided to migrate to [OpenBao](https://github.com/openbao/openbao).

---

This goes back to the question when to define a module, and when use just a package. A module is needed when independent versioning is needed. That is, a piece of software that evolves somewhat independently from others.

The `replace` directive helps with development, but requires some maintenance overhead to avoid breaking things for the clients. The question then becomes whether the benefits are worth the effort.

Frankly, our own repository does not do a good job at that, and it's something that needs to be fixed.

Hope that helps.

**Pull request merged**